lib/IRGen/GenStruct.cpp

```diff
@@ -290,7 +291,46 @@ namespace {
       }
     }
   };
-  
+
+  class ClangRecordAddressOnlyTypeInfo final
+      : public StructTypeInfoBase<ClangRecordAddressOnlyTypeInfo, FixedTypeInfo,
+                                  ClangFieldInfo> {
+    const clang::RecordDecl *ClangDecl;
+
+  public:
+    ClangRecordAddressOnlyTypeInfo(ArrayRef<ClangFieldInfo> fields,
+                                   llvm::Type *storageType, Size size,
+                                   SpareBitVector &&spareBits, Alignment align,
+                                   const clang::RecordDecl *clangDecl)
+        : StructTypeInfoBase(StructTypeInfoKind::ClangRecordAddressOnlyTypeInfo,
+                             fields, storageType, size, std::move(spareBits),
+                             align, IsPOD, IsNotBitwiseTakable, IsFixedSize),
+          ClangDecl(clangDecl) {
+      (void)ClangDecl;
+    }
+
+    TypeLayoutEntry *buildTypeLayoutEntry(IRGenModule &IGM,
+                                          SILType T) const override {
+      return IGM.typeLayoutCache.getOrCreateScalarEntry(*this, T);
+    }
+
+    void initializeFromParams(IRGenFunction &IGF, Explosion &params,
+                              Address addr, SILType T,
+                              bool isOutlined) const override {
+      llvm_unreachable("Cannot initialize address-only type from params.");
```

Comment:
I think we can initialize address-only type from an explosion; the reason why nobody should call this specific method is to avoid violating C++ invariants about an object having a stable address (C++ types that are non-loadable should be created/copied/moved/destroyed only by C++ special member functions). I'd suggest to reword the error to point to that instead.

Reply:
Yep, you're right.